File: SampleVB/RosenBrock.vb
''' <summary>
''' my objective function : inherit objective function
''' </summary>
''' <remarks></remarks>
Public Class RosenBrock : Inherits LibOptimization.Optimization.absObjectiveFunction
    ''' <summary>
    ''' for Rosen block function(Bench mark function)
    ''' </summary>
    ''' <param name="a"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Overrides Function F(ByVal a As List(Of Double)) As Double
        If a Is Nothing Then
            Return 0.0
        End If
        Dim x1 = a(0)
        Dim x2 = a(1)
        Return 100 * (x2 - x1 * x1) * (x2 - x1 * x1) + (1 - x1) * (1 - x1)
    End Function

    Public Overrides Function Gradient(x As List(Of Double)) As List(Of Double)
        Return Nothing
    End Function

    Public Overrides Function Hessian(x As List(Of Double)) As List(Of List(Of Double))
        Return Nothing
    End Function

    Public Overrides Function NumberOfVariable() As Integer
        Return 2
    End Function
End Class

File: SampleCSharp/RosenBrock.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SampleCSharp
{
    /// <summary>
    /// for Rosen block function(Bench mark function)
    /// </summary>
    class RosenBrock : LibOptimization.Optimization.absObjectiveFunction
    {
        private int _dim = 0;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="dim"></param>
        public RosenBrock(int dim)
        {
            if(dim <=1)
            {
                throw new NotImplementedException();
            }
            this._dim = dim;
        }

        public override double F(List<double> x)
        {
            double ret = 0.0;
            for(int i=0;i<_dim-1;i++)
            {
                var temp1 = x[i + 1] - x[i] * x[i];
                var temp2 = x[i] - 1;
                ret += 100 * temp1* temp1 + temp2*temp2;
            }
            return ret;
        }

        public override List<double> Gradient(List<double> aa, double h = 1E-08)
        {
            return null;
        }

        public override List<List<double>> Hessian(List<double> aa, double h = 1E-08)
        {
            return null;
        }

        public override int NumberOfVariable()
        {
            return this._dim;
        }
    }
}
